python/ray/data/_internal/planner/exchange/fast_repartition_task_scheduler.py

```diff
@@ -0,0 +1,87 @@
+from ray.data._internal.stats import StatsDict
+from ray.data.block import Block, BlockMetadata
+from ray.types import ObjectRef
+
+
+class FastRepartitionTaskScheduler(ExchangeTaskScheduler):
+    """
+    The fast (non-shuffle) repartition scheduler.
+
+    First, we calculate global splits needed to produce `output_num_blocks` blocks.
+    After the split blocks are generated accordingly, reduce tasks are scheduled
+    to combine split blocks together.
+    """
+
+    def execute(
+        self,
+        refs: List[RefBundle],
+        output_num_blocks: int,
+        map_ray_remote_args: Optional[Dict[str, Any]] = None,
+        reduce_ray_remote_args: Optional[Dict[str, Any]] = None,
+    ) -> Tuple[List[RefBundle], StatsDict]:
+        input_num_rows = 0
+        input_owned = True
+        for ref_bundle in refs:
+            input_num_rows += ref_bundle.num_rows() or 0
+            if not ref_bundle.owns_blocks:
+                input_owned = False
+
+        # Compute the (output_num_blocks-1) indices needed for
+        # an equal split of the input blocks.
+        indices = []
+        cur_idx = 0
+        for _ in range(output_num_blocks - 1):
+            cur_idx += input_num_rows / output_num_blocks
+            indices.append(int(cur_idx))
+        assert len(indices) < output_num_blocks, (indices, output_num_blocks)
+
+        if map_ray_remote_args is None:
+            map_ray_remote_args = {}
+        if reduce_ray_remote_args is None:
+            reduce_ray_remote_args = {}
+        if "scheduling_strategy" not in reduce_ray_remote_args:
+            reduce_ray_remote_args = reduce_ray_remote_args.copy()
+            reduce_ray_remote_args["scheduling_strategy"] = "SPREAD"
+
+        blocks_with_metadata: List[Tuple[ObjectRef[Block], BlockMetadata]] = []
+        for ref_bundle in refs:
+            blocks_with_metadata.extend(ref_bundle.blocks)
+        if indices:
+            split_return = _split_at_indices(blocks_with_metadata, indices)
+            split_block_refs, split_metadata = [], []
+            for b, m in zip(*split_return):
+                split_block_refs.append(b)
+                split_metadata.extend(m)
+        else:
+            split_block_refs, split_metadata = blocks_with_metadata, []
+
+        reduce_bar = ProgressBar("Fast Repartition", total=output_num_blocks)
+        reduce_task = cached_remote_fn(self._exchange_spec.reduce)
+        reduce_return = [
+            reduce_task.options(**reduce_ray_remote_args, num_returns=2).remote(
+                *self._exchange_spec._reduce_args,
+                *split_block_refs[j],
+            )
+            for j in range(output_num_blocks)
+        ]
+
+        reduce_block_refs, reduce_metadata = zip(*reduce_return)
+        reduce_metadata = reduce_bar.fetch_until_complete(list(reduce_metadata))
+        reduce_bar.close()
+
+        output = []
+        for block, meta in zip(reduce_block_refs, reduce_metadata):
```

Comment:
what about the original empty block handling - https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/fast_repartition.py#L75-L105 ? Do we still care about it? Can you add unit test for it?

Reply:
I had a previous test which I was using to see if this would be an issue, and I incorrectly concluded that there was no impacted. I added more tests which exposed the issue, and added the empty block handling case.